source/opt/loop_unswitch_pass.cpp

```diff
@@ -0,0 +1,1179 @@
+  BlockMapTy old_to_new_bb_;
+  BlockMapTy new_to_old_bb_;
+
+  // Cleans up mapping stats between |loop_| and a cloned loop.
+  void ClearMappingState() {
+    value_map_.clear();
+    old_to_new_bb_.clear();
+    new_to_old_bb_.clear();
+  }
+
+  // Returns the next usable id for the context.
+  uint32_t TakeNextId() { return context_->TakeNextId(); }
+
+  // Removes any block that is tagged as dead, if the block is in
+  // |unreachable_merges| then all block's instructions are replaced by a
+  // OpUnreachable.
+  template <typename Container>
+  void CleanUpCFG(Container* container,
+                  const std::unordered_set<uint32_t>& dead_blocks,
+                  const std::unordered_set<uint32_t>& unreachable_merges) {
+    ir::CFG& cfg = *context_->cfg();
+
+    typename Container::iterator bb_it = container->begin();
+    while (bb_it != container->end()) {
+      ir::BasicBlock* bb = bb_it->get();
+
+      if (unreachable_merges.count(bb->id())) {
+        if (bb->begin() != bb->tail() ||
+            bb->terminator()->opcode() != SpvOpUnreachable) {
+          // Make unreachable, but leave the label.
+          bb->KillAllInsts(false);
+          opt::InstructionBuilder(context_, bb).AddUnreachable();
+          cfg.RemoveNonExistingEdges(bb->id());
+        }
+        ++bb_it;
+      } else if (dead_blocks.count(bb->id())) {
+        cfg.ForgetBlock(bb);
+        // Kill this block.
+        bb->KillAllInsts(true);
+        bb_it = container->erase(bb_it);
+      } else {
+        cfg.RemoveNonExistingEdges(bb->id());
+        ++bb_it;
+      }
+    }
+  }
+
+  // Return true if |c_inst| is a Boolean constant and set |cond_val| with the
+  // value that |c_inst|
+  bool GetConstCondition(const ir::Instruction* c_inst, bool* cond_val) {
+    bool cond_is_const;
+    switch (c_inst->opcode()) {
+      case SpvOpConstantFalse: {
+        *cond_val = false;
+        cond_is_const = true;
+      } break;
+      case SpvOpConstantTrue: {
+        *cond_val = true;
+        cond_is_const = true;
+      } break;
+      default: { cond_is_const = false; } break;
+    }
+    return cond_is_const;
+  }
+
+  // Simplifies |loop| assuming the instruction |to_version_insn| takes the
+  // value |cst_value|.
+  // Requirements:
+  //   - |loop| must be in the LCSSA form;
+  //   - |cst_value| must be constant.
+  // The set |dead_blocks| will contain all the dead basic blocks, and the list
+  // |ordered_loop_bb| will contain the live basic blocks in reverse post-order.
+  void SimplifyLoop(ir::Loop* loop, ir::Instruction* to_version_insn,
+                    ir::Instruction* cst_value,
+                    std::unordered_set<uint32_t>* dead_blocks) {
+    // Version the |loop| body.
+    // Do a DFS that takes into account the specialized value of
+    // to_version_insn. As we go we:
+    //  - Fold instructions;
+    //  - Hoist any loop invariant;
+    //  - Only keep reachable basic block.
```

Comment:
I wondering if it is worth doing all of this work in here.  We could cut down on the amount of code (and potential missed cases), if we just change the value in the branch, and left the rest to other transformations: ccp, simplification, loop invariant code motion, and dead branch elimination.  The will probably need to be run anyway.

Is there a reason we cannot do that?

Reply:
The only real requirement is to kill the constant branch condition (this would create an infinite loop otherwise).

The reason why I made this is to give the pass the opportunity to kill sub-loops and so actively avoid unswitching less obvious dead ones.